src/backend/Donate/Actions/MyDonations/GetDetails/GetMyDonationDetailsService.cs

```diff
@@ -0,0 +1,112 @@
+﻿using Core.Database;
+using Core.Database.Enums;
+using Core.Exceptions;
+using Core.Services;
+using Core.Setup.Auth0;
+using Donate.Actions.DonateForm.GetWarehouses;
+using Donate.Shared;
+using HashidsNet;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+using System.Linq.Dynamic.Core;
+using Z.EntityFramework.Plus;
+
+namespace Donate.Actions.MyDonations.GetDetails
+{
+    public class GetMyDonationDetailsService
+    {
+        private readonly AppDbContext _appDbContext;
+        private readonly ICurrentUserService _currentUserService;
+        private readonly GetCurrentUserService _getCurrentUserService;
+        private readonly RoleService _roleService;
+        private readonly Hashids _hashids;
+        private readonly ILogger<GetWarehousesService> _logger;
+
+        public GetMyDonationDetailsService(
+            AppDbContext appDbContext,
+            ICurrentUserService currentUserService,
+            GetCurrentUserService getCurrentUserService,
+            RoleService roleService,
+            Hashids hashids,
+            ILogger<GetWarehousesService> logger
+            )
+        {
+            _appDbContext = appDbContext;
+            _currentUserService = currentUserService;
+            _getCurrentUserService = getCurrentUserService;
+            _roleService = roleService;
+            _hashids = hashids;
+            _logger = logger;
+        }
+
+        public async Task<GetMyDonationDetailsResponse> GetMyDonationDetails(string id)
+        {
+            var auth0UserInfo = await _currentUserService.GetUserInfo();
+            var currentUser = await _getCurrentUserService.Execute(auth0UserInfo);
+            await _roleService.ThrowIfNoRole(RoleName.Donor, currentUser.Id);
+            var decodedDonationId = _hashids.DecodeSingleLong(id);
+
+
+            var donationById = _appDbContext.Donations.SingleOrDefaultAsync(c => c.Id == decodedDonationId
+               && c.Status == DbEntityStatus.Active || c.Status == DbEntityStatus.Inactive);
+
+            if (donationById == null)
+            {
+                _logger.LogError("Couldn't find donation in database");
+                throw new ApplicationErrorException("Sorry there seems to be a problem with our service");
+            }
+
+            var listOfDonationProducts = await _appDbContext.DonationProducts
+                .Include(c => c.Product)
+                .Where(c => c.Donation == donationById.Result
```

Comment:
To jest całkiem do wywalenia, dołożyłeś sobie roboty. Po prostu tam wyżej zrób
```
_appDbContext.Donations
   .Include(c => c.Products)
       .ThenInclude(c => c.Product).SingleOrDefaultAsync(...)
   .Include(c => c.Warehouse)
     .ThenInclude(c => c.Address)
        .ThenInclude(c => c!.Country)
```
i będziesz miał wszystko za jednym strzałem.

BTW.: W `Where()` nie porównuj obiektów. Pamiętaj, że te pseudo kwerendy są tłumaczone na SQL - używaj np `DonationId`.

Reply:
Gucci geng